hibernate-core/src/main/java/org/hibernate/query/sqm/internal/TypecheckUtil.java

```diff
@@ -360,4 +365,124 @@ public static void assertAssignable(
 			}
 		}
 	}
+
+	public static void assertOperable(SqmExpression<?> left, SqmExpression<?> right, BinaryArithmeticOperator op) {
+		final SqmExpressible<?> leftNodeType = left.getNodeType();
+		final SqmExpressible<?> rightNodeType = right.getNodeType();
+		if ( leftNodeType != null && rightNodeType != null ) {
+			final Class<?> leftJavaType = leftNodeType.getExpressibleJavaType().getJavaTypeClass();
+			final Class<?> rightJavaType = rightNodeType.getExpressibleJavaType().getJavaTypeClass();
+			if ( Number.class.isAssignableFrom( leftJavaType ) ) {
```

Comment:
I think it would be nicer and more extendible/future-proof to add methods on `JavaType` for determining whether some other `JavaType` can be used for a given `BinaryArithmeticOperator`. Just an idea though, we can still do that later.